decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java

```diff
@@ -0,0 +1,305 @@
+        }
+    }
+
+    /**
+     * Index of the start node of next best edge from the slice
+     *
+     * @param sourceSlice
+     * @return index of the node from which next edge starts
+     */
+    private int getNextSourceNodeIndex(Slice sourceSlice) {
+        if (graphEdge == null) {
+            GraphNode sourceNode = sourceSlice.getNextBestNode();
+            return sourceSlice.getNodePosition(sourceNode);
+        } else if (graphEdge.getEdgeColor().equals(EntityColor.RED)) {
+            handleIncomingEdges(graphEdge.getSourceNode(), sourceSlice);
+            return findNextNodeIndexIfCurrentIsRed(sourceSlice);
+        } else {
+            return sourceSlice.getNodePosition(graphEdge.getDestinationNode());
+        }
+
+    }
+
+    /**
+     * We don't have to explore any incoming edges if node is a dead end
+     *
+     * @param sourceNode
+     * @param sourceSlice
+     */
+    private void handleIncomingEdges(GraphNode<T, L> sourceNode, Slice sourceSlice) {
+        if (sourceNode.isDeadEnd() && slices.indexOf(sourceSlice) != slices.size() - 1) {
+            sourceNode.getIncomingEdge().forEach(
+                    edge -> {
+                        edge.setEdgeColor(EntityColor.RED);
+                    }
+            );
+        }
+    }
+
+    /**
+     * Backtracking is required if the green edges in the previous slices are not connected directly or indirectly to the given edge
+     *
+     * @param currentSliceIndex index of the slice from which the edge starts
+     * @param graphEdge         edge of which the predecessors need to verified
+     * @return
+     */
+    public boolean isBacktrackingRequired(int currentSliceIndex, GraphEdge<T, L> graphEdge) {
+        int sliceIndex = currentSliceIndex;
+        for (GraphEdge<T, L> edge = graphEdge; edge != null; --sliceIndex) {
+            edge = edge.getSourceNode().getIncomingEdge()
+                    .stream()
+                    .filter(incomingEdge -> incomingEdge.getEdgeColor().equals(EntityColor.GREEN)).findFirst().orElse(null);
+        }
+        return (sliceIndex != -1);
+
+    }
+
+    /**
+     * Backtracking is required if the green edges in the previous slices are not connected directly or indirectly to the last best edge
+     */
+    private boolean isBacktrackingRequired() {
+        return isBacktrackingRequired(currentSourceSliceIndex, graphEdge);
+    }
+
+
+    /**
+     * Forwardtracking is required if there are green successors to the current edge
+     */
+    private boolean isForwardTrackingRequired() {
+        return (graphEdge.getDestinationNode().getOutgoingEdge()
+                .stream()
+                .filter(outgoingEdge -> outgoingEdge.getEdgeColor().equals(EntityColor.GREEN)).count() != 0);
+    }
+
+
+    /**
+     * if no backtracking or forwardtracking is preformed, then next best edge is is estimated based on start or end node of the last best edge
```

Comment:
Better to be consistent in terminology.  Sometimes it is "backtracking" and sometimes "backwardtracking".